[Profiler] set_class util (part 1 of Record Optimizer)

slgong-fb opened the pull request:
Summary:
Part 1 of Record Optimizer param_groups and states (https://github.com/pytorch/pytorch/pull/84063)
- nnModule and Optimizer have duplicated parts
- create a util function to avoid duplication

Test Plan: buck run mode/opt //caffe2/test:profiler

Differential Revision: D39397210

robieta commented on torch/csrc/autograd/profiler_python.cpp:
```diff
@@ -249,6 +256,28 @@ class ValueCache {
   void trimPrefixes();
 
+  template <CallType C>
+  typename Config<C>::cls_t set_class(
```
This feels strange as part of ValueCache. It seems like there should be a base class for the cache of these specializations.

slgong-fb replied:
updated as a standalone function

robieta commented on torch/csrc/autograd/profiler_python.cpp:
```diff
@@ -249,6 +256,28 @@ class ValueCache {
+      typename Config<C>::cache_t& cache,
+      const typename Config<C>::key_t& key,
+      const typename Config<C>::ephemeral_t& frame,
+      const PyCodeObject* code_match) {
```
If we template get code on CallType then we can make this more generic. E.g. `TORCH_INTERNAL_ASSERT(code.get() == getCode<C>());`

slgong-fb replied:
updated

slgong-fb committed:
[Profiler] set_class util (part 1 of Record Optimizer) (#84779)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84779

Part 1 of Record Optimizer param_groups and states (https://github.com/pytorch/pytorch/pull/84063)
- nnModule and Optimizer have duplicated parts
- create a util function to avoid duplication

Test Plan: buck run mode/opt //caffe2/test:profiler

Differential Revision: D39397210

fbshipit-source-id: 69ff9bd6774367bd23f0048ee1709d47e9504bdc

robieta commented on torch/csrc/autograd/profiler_python.cpp:
```diff
@@ -77,6 +77,14 @@ PyCodeObject* nnModuleCode() {
+PyCodeObject* getCode();
+
+template <>
+PyCodeObject* getCode<CallType::PyModuleCall>() {
```
I don't think we need this indirection. We can just move the body and use `getCode<CallType::PyModuleCall>()` everywhere.

robieta reviewed (approved):
One nit, but overall looks good!
